crates/environ/src/stack_map.rs

```diff
@@ -1,22 +1,92 @@
-use cranelift_bitset::CompoundBitSet;
-use serde_derive::{Deserialize, Serialize};
+use object::{Bytes, LittleEndian, U32Bytes};
+
+struct StackMapSection<'a> {
+    pcs: &'a [U32Bytes<LittleEndian>],
+    pointers_to_stack_map: &'a [U32Bytes<LittleEndian>],
+    stack_map_data: &'a [U32Bytes<LittleEndian>],
+}
+
+impl<'a> StackMapSection<'a> {
+    fn parse(section: &'a [u8]) -> Option<StackMapSection<'a>> {
+        let mut section = Bytes(section);
+        // NB: this matches the encoding written by `append_to` in the
+        // `compile::stack_map` module.
+        let pc_count = section.read::<U32Bytes<LittleEndian>>().ok()?;
+        let pc_count = usize::try_from(pc_count.get(LittleEndian)).ok()?;
+        let (pcs, section) =
+            object::slice_from_bytes::<U32Bytes<LittleEndian>>(section.0, pc_count).ok()?;
+        let (pointers_to_stack_map, section) =
+            object::slice_from_bytes::<U32Bytes<LittleEndian>>(section, pc_count).ok()?;
+        let stack_map_data =
+            object::slice_from_all_bytes::<U32Bytes<LittleEndian>>(section).ok()?;
+        Some(StackMapSection {
+            pcs,
+            pointers_to_stack_map,
+            stack_map_data,
+        })
+    }
+
+    fn lookup(&self, pc: u32) -> Option<StackMap<'a>> {
+        let pc_index = self
+            .pcs
+            .binary_search_by_key(&pc, |v| v.get(LittleEndian))
+            .ok()?;
+        self.get(pc_index)
+    }
+
+    fn into_iter(self) -> impl Iterator<Item = (u32, StackMap<'a>)> + 'a {
+        self.pcs
+            .iter()
+            .enumerate()
+            .map(move |(i, pc)| (pc.get(LittleEndian), self.get(i).unwrap()))
+    }
+
+    /// Returns the stack map corresponding to the `i`th pc.
+    fn get(&self, i: usize) -> Option<StackMap<'a>> {
+        let pointer_to_stack_map = self.pointers_to_stack_map[i].get(LittleEndian) as usize;
+        let data = self.stack_map_data.get(pointer_to_stack_map..)?;
+
+        let (frame_size, data) = data.split_first()?;
+        let (count, data) = data.split_first()?;
+        let data = data.get(..count.get(LittleEndian) as usize)?;
+
+        Some(StackMap {
+            frame_size: frame_size.get(LittleEndian),
+            data,
+        })
+    }
+}
 
 /// A map for determining where live GC references live in a stack frame.
 ///
 /// Note that this is currently primarily documented as cranelift's
 /// `binemit::StackMap`, so for detailed documentation about this please read
 /// the docs over there.
-#[derive(Debug, Serialize, Deserialize)]
-pub struct StackMap {
-    bits: CompoundBitSet,
+pub struct StackMap<'a> {
     frame_size: u32,
+    data: &'a [U32Bytes<LittleEndian>],
```

Comment:
Are we sure we want to re-export `U32Bytes` in the public API? I suspect the answer is "yes" but just wanted to note this.

Reply:
Yeah it's fine here in that `wasmtime-environ` is a "private crate" where we don't consider it part of Wasmtime's public API, just an internal implementation detail. We can probably do more to document that at the crate itself though.